.github/workflows/windows.yml

```diff
@@ -42,52 +42,204 @@ env:
+      - if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) && github.event_name != 'pull_request_target' }}
+        run: git fetch --depth=2 origin $env:GITHUB_SHA
+      - if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) && github.event_name == 'pull_request_target' }}
+        run: git fetch --depth=2 origin refs/pull/${{ github.event_number }}/head
+      - if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+        # Same as `git switch --detach FETCH_HEAD`, but fixes up dirty working
+        # trees, in case the runner image was baked with a dirty working tree.
+        run: |
+          git switch --detach
+          git reset --hard FETCH_HEAD
+
+      - name: ccache
+        # FIXME: “Error: Restoring cache failed: Error: Unable to locate executable file: sh.”
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+        uses: hendrikmuhs/ccache-action@v1.2
+
+      # Install missing tools in a GitHub-hosted runner.
       # Workaround for https://github.com/actions/runner-images/issues/10001:
       - name: Upgrade llvm
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
         run: choco upgrade llvm
-      - name: ccache
-        uses: hendrikmuhs/ccache-action@v1.2
-      - name: Install crown
-        run: cargo install --path support/crown
       - name: Install wixtoolset
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
         run: |
           choco install wixtoolset
           echo "C:\\Program Files (x86)\\WiX Toolset v3.11\\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
       - uses: actions/setup-python@v5
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
         with:
           python-version: "3.10"
       - name: Bootstrap
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
         run: |
           python -m pip install --upgrade pip
           python mach fetch
           python mach bootstrap-gstreamer
       # For some reason WiX isn't currently on the GitHub runner path. This is a
       # temporary workaround until that is fixed.
       - name: Add WiX to Path
+        if: ${{ ! fromJSON(needs.runner-select.outputs.is-self-hosted) }}
         run: |
           "$env:WIX\bin" >> $env:GITHUB_PATH
+
+      # Always install crown, even on self-hosted runners, because it is tightly
+      # coupled to the rustc version, and we may have the wrong version if the
+      # commit we are building uses a different rustc version.
+      - name: Install crown
+        run: cargo install --path support/crown --force
+
+      - name: Check configuration for incremental builds
+        if: ${{ fromJSON(needs.runner-select.outputs.is-self-hosted) }}
+        run: |
+          cat C:\init\incremental_build_debug.txt
+          echo "`$env:LIBCLANG_PATH now = $env:LIBCLANG_PATH"
+          echo "`$env:PATH now = $env:PATH"
+
```

Comment:
This is just logging, for debugging why no incremental builds, right? In that case we might want to rename step to make this more obvious.

Reply:
done in 8935ae67605bb011bf434d284e1310cffdf267d7